### PR TITLE
Improve Neighborhood Loopkup

### PR DESCRIPTION
**Pull request by mfehr**
Replaced the neighborhood logic with lookup tables and removed some of the unused stuff.
The lookup a bit more than twice as fast now, and overall it improves the speed of ESDF generation by 4-7%.

### CODE REVIEW

File: voxblox/include/voxblox/utils/neighbor_tools.h [hunk 1]
```diff
@@ -13,246 +13,87 @@ enum Connectivity : unsigned int {
...
-    for (int j = -1; j <= 1; j += 2) {
-      direction(i) = j;
-      for (int k = -1; k <= 1; k += 2) {
-        direction(next_i) = k;
-        neighbors->emplace_back(global_index + direction);
+  // Get the hierarchical index (block idx, local voxel index) a voxel described
+  // by its hierarchical index and a direction. The main purpose of this
+  // function is to solve the cross-block indexing that happens when looking up
+  // neighbors at the block boundaries.
+  static void getFromHierarchicalIndexAndDirection(
```
Comment (helenol) [inline voxblox/include/voxblox/utils/neighbor_tools.h]: So Hierarchical index is defined as:
```
typedef typename HierarchicalIndexMap::value_type HierarchicalIndex;
```
In block_hash.h, so this is not from HierachicalIndex
Can we call it BlockAndVoxel index as elsewhere? I think this is much clearer and more consistent with other terminology.

File: voxblox/src/integrator/esdf_integrator.cc [hunk 2]
```diff
@@ -294,7 +288,7 @@ void EsdfIntegrator::processRaiseSet() {
   // (2) if the neighbor's parent differs, add it to open (we will have to
   //    update our current distances, of course).
   GlobalIndexVector neighbors;
-
+  Neighborhood<Connectivity::kTwentySix>::IndexMatrix neigbor_indices;
```
Comment (helenol) [inline voxblox/src/integrator/esdf_integrator.cc]: s/neigbor/neighbor

File: voxblox/src/integrator/esdf_integrator.cc [hunk 3]
```diff
@@ -370,12 +366,17 @@ void EsdfIntegrator::processOpenSet() {
...
+        global_index, &neigbor_indices);
 
     // Go through the neighbors and see if we can update any of them.
-    for (const GlobalIndex& neighbor_index : neighbors) {
+    for (unsigned int idx = 0u; idx < neigbor_indices.cols(); ++idx) {
+      const GlobalIndex& neighbor_index = neigbor_indices.col(idx);
+      const SignedIndex& direction =
+          NeighborhoodLookupTables::kOffsets.col(idx);
+      FloatingPoint distance =
+          NeighborhoodLookupTables::kDistances[idx] * voxel_size_;
```
Comment (mfehr) [inline voxblox/src/integrator/esdf_integrator.cc]: If we want to squeeze even a bit more out of this, we could precompute the metric distances based on the voxel size, we could e.g. store it in the esdf integrator as a member when it is initialized.

File: voxblox/include/voxblox/utils/neighbor_tools.h [hunk 4]
```diff
@@ -13,246 +13,87 @@ enum Connectivity : unsigned int {
...
+      VoxelKey& neighbor = neighbors[i];
+      getFromHierarchicalIndexAndDirection(block_index, voxel_index,
+                                           kOffsets.col(i), voxels_per_side,
+                                           &neighbor.first, &neighbor.second);
     }
   }
-}
-
-template <typename VoxelType>
-SignedIndex NeighborTools<VoxelType>::getOffsetBetweenVoxels(
```
Comment (helenol) [inline voxblox/include/voxblox/utils/neighbor_tools.h]: I use this in the skeleton generation and planning. Please don't remove this. :( :( :( :( : (

File: voxblox/include/voxblox/utils/neighbor_tools.h (hunk 1 above)
Comment (mfehr) [reply voxblox/include/voxblox/utils/neighbor_tools.h]: will do

**Comment by mfehr** [comment]
@helenol ptal

File: voxblox/include/voxblox/utils/neighbor_tools.h [hunk 5]
```diff
@@ -13,246 +13,97 @@ enum Connectivity : unsigned int {
...
-    // This can be negative, but that's totally fine.
-    voxel_offset(i) += end_voxel_index(i) - start_voxel_index(i);
+  // Get the signed offset between the global indices of two voxels.
+  static SignedIndex getOffsetBetweenVoxels(const BlockIndex& start_block_index,
+                                            const VoxelIndex& start_voxel_index,
+                                            const BlockIndex& end_block_index,
+                                            const VoxelIndex& end_voxel_index,
+                                            const size_t voxels_per_side) {
+    CHECK_NE(voxels_per_side, 0u);
+    return (end_voxel_index - start_voxel_index) +
```
Comment (helenol) [inline voxblox/include/voxblox/utils/neighbor_tools.h]: hahahahahahahahahahahahahahahahahaha I'm great.

**Pull request merged**